drivers/wifi/siwx91x/siwx91x_wifi.c

```diff
@@ -0,0 +1,368 @@
+	}
+
+	siwx91x_on_join_ipv4(sidev);
+	siwx91x_on_join_ipv6(sidev);
+
+	return 0;
+}
+
+static int siwx91x_connect(const struct device *dev, struct wifi_connect_req_params *params)
+{
+	sl_wifi_client_configuration_t wifi_config = {
+		.bss_type = SL_WIFI_BSS_TYPE_INFRASTRUCTURE,
+	};
+	int ret;
+
+	switch (params->security) {
+	case WIFI_SECURITY_TYPE_NONE:
+		wifi_config.security = SL_WIFI_OPEN;
+		wifi_config.encryption = SL_WIFI_NO_ENCRYPTION;
+		break;
+	case WIFI_SECURITY_TYPE_WPA_PSK:
+		wifi_config.security = SL_WIFI_WPA;
+		wifi_config.encryption = SL_WIFI_DEFAULT_ENCRYPTION;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	case WIFI_SECURITY_TYPE_PSK:
+		wifi_config.security = SL_WIFI_WPA2;
+		wifi_config.encryption = SL_WIFI_TKIP_ENCRYPTION;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	case WIFI_SECURITY_TYPE_PSK_SHA256:
+		wifi_config.security = SL_WIFI_WPA2;
+		wifi_config.encryption = SL_WIFI_CCMP_ENCRYPTION;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	case WIFI_SECURITY_TYPE_SAE:
+		/* TODO: Support the case where MFP is not required */
+		wifi_config.security = SL_WIFI_WPA3;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	case WIFI_SECURITY_TYPE_WPA_AUTO_PERSONAL:
+		wifi_config.security = SL_WIFI_WPA2;
+		wifi_config.encryption = SL_WIFI_DEFAULT_ENCRYPTION;
+		wifi_config.credential_id = SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID;
+		break;
+	/* Zephyr WiFi shell doesn't specify how to pass credential for these
+	 * key managements.
+	 */
+	case WIFI_SECURITY_TYPE_WEP: /* SL_WIFI_WEP/SL_WIFI_WEP_ENCRYPTION */
+	case WIFI_SECURITY_TYPE_EAP: /* SL_WIFI_WPA2_ENTERPRISE/<various> */
+	case WIFI_SECURITY_TYPE_WAPI:
+	default:
+		return -ENOTSUP;
+	}
+
+	if (params->band != WIFI_FREQ_BAND_UNKNOWN && params->band != WIFI_FREQ_BAND_2_4_GHZ) {
+		return -ENOTSUP;
+	}
+
+	if (params->psk_length) {
+		sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID, SL_NET_WIFI_PSK,
+				      params->psk, params->psk_length);
+	}
+
+	if (params->sae_password_length) {
+		sl_net_set_credential(SL_NET_DEFAULT_WIFI_CLIENT_CREDENTIAL_ID, SL_NET_WIFI_PSK,
+				      params->sae_password, params->sae_password_length);
+	}
+
+	if (params->channel != WIFI_CHANNEL_ANY) {
+		wifi_config.channel.channel = params->channel;
+	}
+
+	wifi_config.ssid.length = params->ssid_length,
+	memcpy(wifi_config.ssid.value, params->ssid, params->ssid_length);
+
+	ret = sl_wifi_connect(SL_WIFI_CLIENT_INTERFACE, &wifi_config, 0);
+	if (ret != SL_STATUS_IN_PROGRESS) {
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int siwx91x_disconnect(const struct device *dev)
+{
+	struct siwx91x_dev *sidev = dev->data;
+	int ret;
+
+	ret = sl_wifi_disconnect(SL_WIFI_CLIENT_INTERFACE);
+	if (ret) {
+		return -EIO;
+	}
+	if (IS_ENABLED(CONFIG_WIFI_SILABS_SIWX91X_NET_STACK_NATIVE)) {
+		net_eth_carrier_off(sidev->iface);
```

Comment:
Other wifi drivers seem to call `net_if_dormant_on()` in this scenario. I suppose carrier can be turned off, just wondering did you consider using the dormant state?

Reply:
Replaced `net_eth_carrier_off()` by `net_if_dormant_on()`.